deepchem/models/torch_models/gan.py

```diff
@@ -454,3 +454,326 @@ def _list_or_tensor(inputs):
+
+        Subclasses must override this to construct the generator.  The returned
+        value should be a tf.keras.Model whose inputs are a batch of noise, followed
+        by any conditional inputs.  The number and shapes of its outputs must match
+        the return value from get_data_input_shapes(), since generated data must
+        have the same form as training data.
+        """
+        raise NotImplementedError("Subclasses must implement this.")
+
+    def create_discriminator(self):
+        """Create and return a discriminator.
+
+        Subclasses must override this to construct the discriminator.  The returned
+        value should be a tf.keras.Model whose inputs are all data inputs, followed
+        by any conditional inputs.  Its output should be a one dimensional tensor
+        containing the probability of each sample being a training sample.
+        """
+        raise NotImplementedError("Subclasses must implement this.")
+
+    def fit_gan(self,
+                batches,
+                generator_steps=1,
+                max_checkpoints_to_keep=5,
+                checkpoint_interval=1000,
+                restore=False):
+        """Train this model on data.
+
+        Parameters
+        ----------
+        batches: iterable
+            batches of data to train the discriminator on, each represented as a dict
+            that maps Inputs to values.  It should specify values for all members of
+            data_inputs and conditional_inputs.
+        generator_steps: float
+            the number of training steps to perform for the generator for each batch.
+            This can be used to adjust the ratio of training steps for the generator
+            and discriminator.  For example, 2.0 will perform two training steps for
+            every batch, while 0.5 will only perform one training step for every two
+            batches.
+        max_checkpoints_to_keep: int
+            the maximum number of checkpoints to keep.  Older checkpoints are discarded.
+        checkpoint_interval: int
+            the frequency at which to write checkpoints, measured in batches.  Set
+            this to 0 to disable automatic checkpointing.
+        restore: bool
+            if True, restore the model from the most recent checkpoint before training
+            it.
+        """
+        self._ensure_built()
+
+        gen_train_fraction = 0.0
+        discrim_error = 0.0
+        gen_error = 0.0
+        discrim_average_steps = 0
+        gen_average_steps = 0
+        time1 = time.time()
+
+        for feed_dict in batches:
+            # Every call to fit_generator() will increment global_step, but we only
+            # want it to get incremented once for the entire batch, so record the
+            # value and keep resetting it.
+
+            global_step = self.get_global_step()
+
+            # Train the discriminator.
+
+            inputs = [self.get_noise_batch(self.batch_size)]
+
+            temp_data = []
```

Comment:
Don't use temp_ in variable names. Try something more descriptive

Reply:
fixed